rust_src/src/lread.rs

```diff
@@ -404,4 +405,52 @@ pub fn read_char_exclusive(
     }
 }
 
+/// Delete the symbol named NAME, if any, from OBARRAY.
+/// The value is t if a symbol was found and deleted, nil otherwise.
+/// NAME may be a string or a symbol.  If it is a symbol, that symbol
+/// is deleted, if it belongs to OBARRAY--no other symbol is deleted.
+/// OBARRAY, if nil, defaults to the value of the variable `obarray'.
+/// usage: (unintern NAME OBARRAY)
+#[lisp_fn(min = "1")]
+pub fn unintern(name: LispSymbolOrString, obarray: Option<LispObarrayRef>) -> LispObject {
+    let obarray = obarray.unwrap_or_else(LispObarrayRef::global);
+    let obarray: LispObarrayRef = check_obarray(obarray.into()).into();
+
+    let tem = obarray.lookup(name);
+    if tem.is_integer() {
+        return Qnil;
+    }
+    // If arg was a symbol, don't delete anything but that symbol itself.
+    if name.is_symbol() && name != tem {
+        return Qnil;
+    }
+
+    let mut temp: LispSymbolRef = tem.into();
+    temp.set_uninterned();
+
+    let hash = unsafe { oblookup_last_bucket_number };
+
+    let mut obarray = LispObject::from(obarray).as_vector_or_error();
+    let symbol: LispSymbolRef = obarray.get(hash).into();
+
+    if symbol == temp {
+        match symbol.get_next() {
```

Comment:
`get_next` is the same as `iter().next()` I am not certain we need it.

Reply:
Due to how we have the Iterator trait set up for `LispSymbolIter`, `iter().next()` returns the current item (which makes sense from a Rust perspective, not so much from C). ironically, here we'd need to do another `.next()` on the iterator to _really_ get the next item. I adjusted the code to `iter().nth(1)` which, admittedly, is still awkward. And I left the `get_next` function in here for symmetry with `set_next`, seems a bit cleaner. Open to other options (no pun intended) ...